paddle/fluid/framework/executor_cache.h

```diff
@@ -45,121 +45,90 @@ void ParseSafeEagerDeletionSkipVars(
-    auto key_val = key_hash_func_(key);
+  const BuildStrategy& GetBuildStrategy(int64_t program_id) {
+    // If not found, insert build_strategy with default value.
+    return strategy_map_[program_id];
+  }
+
+  void SetBuildStrategy(int64_t program_id,
+                        const BuildStrategy& build_strategy) {
     PADDLE_ENFORCE_EQ(
-        Has(key_val), true,
-        platform::errors::NotFound("%s doesn't exist in ExecutorInfoCache",
-                                   key.DebugString()));
-    return info_map_[key_val];
+        strategy_map_.count(program_id), 0,
+        platform::errors::PreconditionNotMet(
+            "program_id: %s already exist in ExecutorInfoCache", program_id));
+    strategy_map_[program_id] = build_strategy;
   }
 
-  bool Has(const CacheKey& key) const {
-    auto key_val = key_hash_func_(key);
-    return Has(key_val);
+  bool Has(int64_t program_id, bool is_grad) {
+    return info_map_.find(program_id) != info_map_.end() &&
+           info_map_[program_id].IsAvailable(is_grad);
   }
 
-  bool Has(const KeyType& key) const {
-    return info_map_.find(key) != info_map_.end();
+  ExecutorInfo::CacheValue& GetMutable(int64_t program_id, bool is_grad) {
+    return info_map_[program_id].GetMutable(is_grad);
   }
 
-  void Insert(const CacheKey& key, ValueType value) {
-    auto key_val = key_hash_func_(key);
-    PADDLE_ENFORCE_EQ(
-        Has(key_val), false,
-        platform::errors::NotFound("%s has existed in ExecutorInfoCache",
-                                   key.DebugString()));
-    info_map_.insert({key_val, value});
+  void UpdateSkipEagerDeleteVars(int64_t program_id, bool is_grad,
+                                 const std::vector<std::string>& skip_vars) {
+    auto& cached_value = GetMutable(program_id, is_grad);
+    cached_value.skip_eager_delete_vars_ = std::move(skip_vars);
   }
 
-  size_t Size() const { return info_map_.size(); }
+  std::vector<std::string>& SkipEagerDeleteVars(int64_t program_id,
+                                                bool is_grad) {
+    auto& cached_value = GetMutable(program_id, is_grad);
+    return cached_value.skip_eager_delete_vars_;
+  }
 
-  void Finalize();
+  void Finalize() {
+    // NOTE(Aurelius84): DO NOT perform finalize in destructor
+    // to avoid problems caused by destructor order of static
+    // object.
+    info_map_.clear();
+    strategy_map_.clear();
+  }
 
  private:
-  ExecutorInfoCache() = default;
-  DISABLE_COPY_AND_ASSIGN(ExecutorInfoCache);
-
-  KeyHasher key_hash_func_;
-  std::unordered_map<KeyType, ValueType> info_map_;
+  std::unordered_map<int64_t, ExecutorInfo> info_map_;
+  std::unordered_map<int64_t, BuildStrategy> strategy_map_;
 };
 
 using CacheInfo =
     std::pair<std::shared_ptr<ParallelExecutor>, bool /*is_new_created*/>;
 
-CacheInfo GetExecutorInfoFromCache(const ExecutorInfoCache::CacheKey& cache_key,
+CacheInfo GetExecutorInfoFromCache(const ProgramDesc* program_desc,
```

Comment:
Is `program_desc` possible to be a `nullptr`? If is not an optional parameter, coding style suggest to replace it by const reference.
https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

Reply:
Thanks, done